refactor(api, hardware): ot3: Handle move position response

sfoster1 opened the pull request:
After https://github.com/Opentrons/ot3-firmware/pull/304 , the firmware will return the current value of the position tracker in micrometers rather than steps, which means we can use its value to estimate the current position of the system. Add handling for the position value in MoveCompleted messages from Canbus and bubble it up to the position tracker in the controller.

sfoster1 committed:
feat(hardware): return positions from moves

The MoveGroupRunner can now accumulate and return the current position
elements (which are now in micrometers, and are transformed to
millimeters in the process) from the remote hardware, providing a
consistent position estimate element to the python side of things.

sfoster1 committed:
refactor(api): use move position responses

The move group runner now returns the position from the firmware after
it moves. Use that value set to update position rather than the
projected position based on the move group math.

amitlissack reviewed (approved):
Looks nice to me

sfoster1 committed:
chore(hardware): add integration tests for moves

sfoster1 committed:
fix integration tests

DerekMaggio commented on hardware/tests/firmware_integration/test_move_groups.py:
```diff
@@ -80,3 +83,101 @@ async def test_add_moves(
+    all_motor_node_step_sizes: Dict[NodeId, float],
+) -> None:
+    """Test that moving returns correct positions."""
+    # First, we should test homing. We have to home because the emulator stack is
+    # long-lived beyond each test, so we might as well make sure we're homing
+    # from a non-zero position
+    prep_move = [
+        create_step(
+            distance={motor_node: motor_node.value for motor_node in all_motor_nodes},
+            velocity={
+                motor_node: motor_node.value / 10 for motor_node in all_motor_nodes
+            },
+            acceleration={motor_node: 0 for motor_node in all_motor_nodes},
+            duration=np.float64(1),
+            present_nodes=all_motor_nodes,
+        )
+    ]
+    prep_runner = MoveGroupRunner([prep_move])
+    # throw away this position - it may change depending on previous state of
+    # simulators
+    await prep_runner.run(can_messenger)
```
Are you running this before home to make sure the robot is a state other than home before testing homing?

sfoster1 replied:
Yeah

sfoster1 replied:
Yeah, that's correct. We have to home anyway so we might as well simultaneously test it

DerekMaggio commented on hardware/tests/opentrons_hardware/hardware_control/test_move_group_runner.py:
```diff
@@ -479,6 +483,8 @@ async def test_single_move(
         ]
     )
+    assert len(position) == 1
+    assert position[0][1].payload.current_position_um.value == 246000
```
Threw me off for a second why we were expecting it to be 246000. 
Realized that it was the distance converted from mm to um. 
Pulling these scalars out to constants such as `EXPECTED_HEAD_MICROMETERS` would make this instantly clear

sfoster1 replied:
that's what the `um` is though. i'm usually all for putting constants in variables with names, but in the case of literal comparisons in unit tests I personally get way more value out of seeing the number right there.

sfoster1 replied:
I guess... I like named values for constants in code, but when doing immediate comparisons in tests I'm a much bigger fan of just having the number there to look at. Plus the unit's in the left hand variable name!

DerekMaggio reviewed (approved):
Looks good to me. Just a question and a suggestion

sfoster1 committed:
fix home present-nodes limitation

This was causing exceptions because the present-axis checks were
mismatched and would present the move group runner with empty move
groups which it does NOT want to deal with. Avoid this with a new test
that makes sure that home doesn't generate empty move groups.

sfoster1 committed:
fix(hardware): Move group handles more empty moves

The move groups are a data structure that can be "empty" on many
different levels; make sure we handle all of them.
